common/quotas/clocked_rate_limiter.go

```diff
@@ -131,12 +133,34 @@ func (l ClockedRateLimiter) WaitN(ctx context.Context, token int) error {
 		close(waitExpired)
 	})
 	defer timer.Stop()
-	select {
-	case <-ctx.Done():
-		reservation.Cancel()
-		return fmt.Errorf("%w: %v", ErrRateLimiterWaitInterrupted, ctx.Err())
-	case <-waitExpired:
-		return nil
+
+	for {
+		select {
+		case <-ctx.Done():
+			reservation.Cancel()
+			return fmt.Errorf("%w: %v", ErrRateLimiterWaitInterrupted, ctx.Err())
+		case <-waitExpired:
+			return nil
+		case <-l.recycleCh:
+			if token > 1 {
+				break // recycling 1 token to a process requesting >1 tokens is a no-op
+			}
+
+			// Cancel() reverses the effects of this Reservation on the rate limit as much as possible,
+			// considering that other reservations may have already been made. Normally, Cancel() indicates
+			// that the reservation holder will not perform the reserved action, so it would make the most
+			// sense to cancel the reservation whose token was just recycled. However, we don't have access
+			// to the recycled reservation anymore, so instead we cancel the current reservation as a proxy.
```

Comment:
even if we did have access to it, Cancel() would be a no-op on that one since it's already past its time...
```suggestion
			// to the recycled reservation anymore, and even if we did, Cancel on a reservation that
			// has fully waited is a no-op, so instead we cancel the current reservation as a proxy.
```

Reply:
you're right, thanks for catching